File: packages/just_audio_edmm/just_audio/android/src/main/java/com/ryanheise/just_audio/BetterEventChannel.java
package com.ryanheise.just_audio;

import io.flutter.plugin.common.BinaryMessenger;
import io.flutter.plugin.common.EventChannel;
import io.flutter.plugin.common.EventChannel.EventSink;

public class BetterEventChannel implements EventSink {
    private EventSink eventSink;
    private final Runnable onListen;
    private final Runnable onCancel;

	public BetterEventChannel(final BinaryMessenger messenger, final String id) {
        this(messenger, id, null, null);
    }

	public BetterEventChannel(
            final BinaryMessenger messenger,
            final String id,
            final Runnable onListen,
            final Runnable onCancel) {
        this.onListen = onListen;
        this.onCancel = onCancel;
        EventChannel eventChannel = new EventChannel(messenger, id);
        eventChannel.setStreamHandler(new EventChannel.StreamHandler() {
            @Override
            public void onListen(final Object arguments, final EventSink eventSink) {
                BetterEventChannel.this.eventSink = eventSink;
                if (BetterEventChannel.this.onListen != null) {
                    BetterEventChannel.this.onListen.run();
                }
            }

            @Override
            public void onCancel(final Object arguments) {
                if (BetterEventChannel.this.onCancel != null) {
                    BetterEventChannel.this.onCancel.run();
                }
                eventSink = null;
            }
        });
	}

    @Override
    public void success(Object event) {
        if (eventSink != null) eventSink.success(event);
    }

    @Override
    public void error(String errorCode, String errorMessage, Object errorDetails) {
        if (eventSink != null) eventSink.error(errorCode, errorMessage, errorDetails);
    }

    @Override
    public void endOfStream() {
        if (eventSink != null) eventSink.endOfStream();
    }
}

File: packages/just_audio_edmm/just_audio/darwin/just_audio/Sources/just_audio/BetterEventChannel.m
#import "./include/just_audio/BetterEventChannel.h"

@implementation BetterEventChannel {
    FlutterEventChannel *_eventChannel;
    FlutterEventSink _eventSink;
    void (^_onListen)(void);
    void (^_onCancel)(void);
}

- (instancetype)initWithName:(NSString*)name messenger:(NSObject<FlutterBinaryMessenger> *)messenger {
    return [self initWithName:name messenger:messenger onListen:nil onCancel:nil];
}

- (instancetype)initWithName:(NSString*)name
                   messenger:(NSObject<FlutterBinaryMessenger> *)messenger
                    onListen:(void (^)(void))onListen
                    onCancel:(void (^)(void))onCancel {
    self = [super init];
    NSAssert(self, @"super init cannot be nil");
    _eventChannel =
        [FlutterEventChannel eventChannelWithName:name binaryMessenger:messenger];
    [_eventChannel setStreamHandler:self];
    _eventSink = nil;
    _onListen = [onListen copy];
    _onCancel = [onCancel copy];
    return self;
}

- (FlutterError*)onListenWithArguments:(id)arguments eventSink:(FlutterEventSink)eventSink {
    _eventSink = eventSink;
    if (_onListen) _onListen();
    return nil;
}

- (FlutterError*)onCancelWithArguments:(id)arguments {
    if (_onCancel) _onCancel();
    _eventSink = nil;
    return nil;
}

- (void)sendEvent:(id)event {
    if (!_eventSink) return;
    _eventSink(event);
}

- (void)dispose {
    if (_eventSink) {
        @try {
            _eventSink(FlutterEndOfEventStream);
        } @catch (NSException *exception) {
            NSLog(@"Exception while ending event stream: %@", exception.reason);
        }
    }
    [_eventChannel setStreamHandler:nil];
}

@end
